models/src/main/scala/coop/rchain/models/EqualsM.scala

```diff
@@ -0,0 +1,159 @@
+  * The generative tests should catch that, provided the offending case class is used transitively.
+  *
+  * @tparam A
+  */
+trait EqualM[A] {
+
+  def equal[F[_]: Sync](self: A, other: A): F[Boolean]
+
+  def equals[F[_]: Sync](a: A, b: Any): F[Boolean] =
+    if (a.getClass.isInstance(b)) {
+      equal[F](a, a.getClass.cast(b))
+    } else false.pure[F]
+
+}
+
+object EqualM extends EqualMDerivation {
+
+  def apply[A](implicit ev: EqualM[A]): EqualM[A] = ev
+
+  def by[A, B: EqualM](f: A => B): EqualM[A] = new EqualM[A] {
+
+    override def equal[F[_]: Sync](self: A, other: A): F[Boolean] =
+      EqualM[B].equal(f(self), f(other))
+
+  }
+
+  /**
+    * EqualM instance delegating to Any#equa(). Hence the type bound.
+    * This MUST NOT be used for types having a recursive structure
+    * nor for types whose EqualCode() delegates to others.
+    * Otherwise the stacksafety provided by the EqualM typeclass is broken.
+    *
+    * @tparam A
+    * @return
+    */
+  def opaqueEqual[A]: EqualM[A] = new EqualM[A] {
+    override def equal[F[_]: Sync](self: A, other: A): F[Boolean] = Sync[F].pure(self == other)
+  }
+
+  implicit val IntEqual: EqualM[Int]                           = opaqueEqual
+  implicit val FloatEqual: EqualM[Float]                       = opaqueEqual
+  implicit val LongEqual: EqualM[Long]                         = opaqueEqual
+  implicit val DoubleEqual: EqualM[Double]                     = opaqueEqual
+  implicit val StringEqual: EqualM[String]                     = opaqueEqual
+  implicit val BooleanEqual: EqualM[Boolean]                   = opaqueEqual
+  implicit val BitSetEqual: EqualM[BitSet]                     = opaqueEqual
+  implicit val ByteStringEqual: EqualM[ByteString]             = opaqueEqual
+  implicit val Blake2b512RandomEqual: EqualM[Blake2b512Random] = opaqueEqual
+  implicit def alwaysEqualEqual[A]: EqualM[AlwaysEqual[A]]     = opaqueEqual
+
+  implicit def seqEqual[A: EqualM]: EqualM[Seq[A]] = new EqualM[Seq[A]] {
+
+    override def equal[F[_]: Sync](self: Seq[A], other: Seq[A]): F[Boolean] = {
+      val pairs = self.toStream.zip(other)
+      Sync[F].delay(self.length == other.length) <&&>
+        pairs.forallM(tupled(EqualM[A].equal[F]))
+    }
+
+  }
+
+  implicit def coevalEqual[A: EqualM]: EqualM[Coeval[A]] = by(_.value)
+
+  implicit val ParEqual: EqualM[Par] = gen[Par]
+  implicit val ExprEqual             = gen[Expr]
+  implicit val VarEqual              = gen[Var]
+  implicit val SendEqual             = gen[Send]
+  implicit val ReceiveEqual          = gen[Receive]
+  implicit val ReceiveBindEqual      = gen[ReceiveBind]
+  implicit val NewEqual              = gen[New]
+  implicit val MatchEqual            = gen[Match]
+
+  implicit val ConnectiveEqual = gen[Connective]
+
+  implicit val ESetEqual = gen[ESet]
+  implicit val EMapEqual = gen[EMap]
+
+  implicit val SortedParHashSetEqual: EqualM[SortedParHashSet] = by(_.sortedPars)
```

Comment:
Why not do `gen[SortedParHashSet]`? Now we have the logic of how to do `equal` for `SortedParHashSet` in two places. Same goes for `SortedParMapEqual`.

Reply:
`gen` is only doing what we should be doing in the first place - implementing equals/hashCode using all the fields. We have a custom `equals` for those types though, and thus the corresponding typeclass instances need to be custom too :)